From 09dcec0ea5a2fadd1783f408036c478d7fef6ee4 Mon Sep 17 00:00:00 2001
From: Radu Solea <radu.solea@nxp.com>
Date: Wed, 8 Mar 2017 16:34:15 +0200
Subject: [PATCH 1562/5242] Add missing NULL checks in CAAM sm

commit  eb31c0531ebad057ec43dc364d0d7493a80fcbb3 from
https://source.codeaurora.org/external/imx/linux-imx.git

    Missing NULL checks in CAAM sm_store and sm_test cause kernel
    crashes if caam init fails.

Signed-off-by: Radu Solea <radu.solea@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/crypto/caam/sm_store.c |    8 +++++++-
 drivers/crypto/caam/sm_test.c  |    9 ++++++++-
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/drivers/crypto/caam/sm_store.c b/drivers/crypto/caam/sm_store.c
index 08aba66..96632b6 100644
--- a/drivers/crypto/caam/sm_store.c
+++ b/drivers/crypto/caam/sm_store.c
@@ -1,4 +1,3 @@
-
 /*
  * CAAM Secure Memory Storage Interface
  * Copyright (C) 2008-2015 Freescale Semiconductor, Inc.
@@ -1001,6 +1000,13 @@ int caam_sm_startup(struct platform_device *pdev)
 	ctrlpriv = dev_get_drvdata(ctrldev);
 
 	/*
+	 * If ctrlpriv is NULL, it's probably because the caam driver wasn't
+	 * properly initialized (e.g. RNG4 init failed). Thus, bail out here.
+	 */
+	if (!ctrlpriv)
+		return -ENODEV;
+
+	/*
 	 * Set up the private block for secure memory
 	 * Only one instance is possible
 	 */
diff --git a/drivers/crypto/caam/sm_test.c b/drivers/crypto/caam/sm_test.c
index 841e300..99b1421 100644
--- a/drivers/crypto/caam/sm_test.c
+++ b/drivers/crypto/caam/sm_test.c
@@ -1,4 +1,3 @@
-
 /*
  * Secure Memory / Keystore Exemplification Module
  * Copyright (C) 2012-2015 Freescale Semiconductor, Inc. All Rights Reserved
@@ -106,6 +105,14 @@ int caam_sm_example_init(struct platform_device *pdev)
 	 */
 	ctrldev = &pdev->dev;
 	ctrlpriv = dev_get_drvdata(ctrldev);
+
+	/*
+	 * If ctrlpriv is NULL, it's probably because the caam driver wasn't
+	 * properly initialized (e.g. RNG4 init failed). Thus, bail out here.
+	 */
+	if (!ctrlpriv)
+		return -ENODEV;
+
 	ksdev = ctrlpriv->smdev;
 	kspriv = dev_get_drvdata(ksdev);
 	if (kspriv == NULL)
-- 
1.7.9.5

